### PR TITLE
Added fast exponentation [C]

### LINKED ISSUE #66: Exponentation by squaring [C]
https://en.wikipedia.org/wiki/Exponentiation_by_squaring

It is very much efficient algorithm to calculate x raised to power k.

Edit: time complexity is O( (nlogx) ^ k)

### PR DESCRIPTION
**Pull request by Monal5031**
Fixes #66 

#### By submitting this pull request I confirm I've read and complied with the below requirements.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm name} [{Language}]`, not `Update README.md` or `Added new code`.

### CODE REVIEW

File: exponentation_by_squaring/exponentation_by_squaring.c [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include<stdio.h>
+
+/* Algorithm to calculate x raised to power k very efficiently
+*  Can be used as a great alternative to normal pow function
+*  base - base of the expression
+*  power - power of expression
+*/
+
+long int exponentation_by_squaring(int base, int power) {
```
Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: The base can be non-integer for the case I guess. Also in that case, we should go with `long double` or at least `double` for the return type of the function.

Comment (Monal5031) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Double will be required when we need to calculate for -ve powers, in this case i have considered for +ve powers only

Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: WIll this definition be good?
```C
long double exponentation_by_squaring(long double, int);
```
In this way, we can easily pass integer and float type number as first parameter and get result as `long double`, which can be casted to desired type.

Comment (Monal5031) [inline exponentation_by_squaring/exponentation_by_squaring.c]: @singhpratyush Yes, it would be actually good to have this definition.

Comment (Monal5031) [inline exponentation_by_squaring/exponentation_by_squaring.c]: @singhpratyush The return type `long double` is not giving answer for any range of inputs I tested but making it double works for negative and positive both the cases.

Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: > not giving answer

What type of answer?

Comment (Monal5031) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Whatever input I give the answer always comes out to be 0.00

Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: I don't see an issue with my code - 
```c
long double exponentation_by_squaring(long double base, unsigned int power) {
    if (power == 0)     // Base case1
        return 1;
    else if (power == 1)    // Base case2
        return base;
    else if (power % 2 == 0)
        return exponentation_by_squaring(base * base, power / 2);
    else if (power % 2 == 1)
        return exponentation_by_squaring(base * base, (power - 1) / 2);
}

int main() {
    long double base = 2;
    unsigned int power = 16;
    printf("%LG raised to %d is %LG\n", base, power, exponentation_by_squaring(base, power));
    return 0;
}
```

Comment (Monal5031) [inline exponentation_by_squaring/exponentation_by_squaring.c]: I guess maybe its some problem with my system.
It seems to work fine on online compiler's too.

Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: I guess you are missing the power of `GCC`.

File: exponentation_by_squaring/exponentation_by_squaring.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#include<stdio.h>
+
+/* Algorithm to calculate x raised to power k very efficiently
+*  Can be used as a great alternative to normal pow function
+*  base - base of the expression
+*  power - power of expression
+*/
+
+long double exponentation_by_squaring(double base, int power) {
+    if (power < 0)		// Negative power case
```
Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Use spaces or tabs consistently.

File: exponentation_by_squaring/exponentation_by_squaring.c [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        return exponentation_by_squaring(1 / base, -power);
+    else if (power == 0)	// Base case1
+        return 1;
+    else if (power == 1)	// Base case2
+        return base;
+    else if (power % 2 == 0)
+        return exponentation_by_squaring(base * base, power / 2);
+    else if (power % 2 == 1)
+        return exponentation_by_squaring(base * base, (power - 1) / 2);
```
Comment (aviaryan) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Logic seems to be wrong here.
Assume base=2, power=3. 
This code will give 4. 
You should do `return base * exponentation_by_squaring(base * base, (power - 1) / 2);`.

File: exponentation_by_squaring/exponentation_by_squaring.c [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+int main() {
+    long double base = 2;
+    int power = -1;
```
Comment (aviaryan) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Put a better example here. 
Negative powers is not why people will frequently use this function for. 
If you did this to show that the function can be used for negative powers, better add it as a comment.

File: exponentation_by_squaring/exponentation_by_squaring.c (hunk 3 above)
Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Damn I missed it.

File: exponentation_by_squaring/exponentation_by_squaring.c (hunk 4 above)
Comment (Monal5031) [inline exponentation_by_squaring/exponentation_by_squaring.c]: @aviaryan I used negative power to check if it was working properly, I'll update it with some high value.

Comment (singhpratyush) [inline exponentation_by_squaring/exponentation_by_squaring.c]: Better add a case for both positive and negative power.

**Comment by Monal5031** [comment]
@aviaryan Please review.

File: README.md [hunk 5]
```diff
@@ -15,9 +15,14 @@ Community (college) maintained list of Algorithms and Data Structures implementa
 | [Binary Search](https://en.wikipedia.org/wiki/Binary_search_algorithm) | [\[X\]](binary_search/binary_search.c) | | [\[X\]](binary_search/BinarySearch.java) | |
 | [Coin Change Problem](http://www.algorithmist.com/index.php/Coin_Change) | [\[X\]](coin_change_problem/coin_change_problem.c) | | | |
 | [Euclidean GCD](https://en.wikipedia.org/wiki/Euclidean_algorithm) | [\[X\]](euclidean_gcd/euclidean_gcd.c) | | [\[X\]](euclidean_gcd/EuclideanGCD.java) | [\[X\]](euclidean_gcd/euclidean_gcd.py) |
+<<<<<<< HEAD
```
Comment (jsroyal) [inline README.md]: resolve the conflict

**Comment by Monal5031** [comment]
@jsroyal @aviaryan Please review

File: README.md (hunk 5 above)
Comment (Monal5031) [inline README.md]: Fixed.

**Pull request merged**